IOMC/ParticleGuns/src/BeamMomentumGunProducer.cc

```diff
@@ -0,0 +1,154 @@
+#include <ostream>
+#include "IOMC/ParticleGuns/interface/BeamMomentumGunProducer.h"
+
+#include "SimDataFormats/GeneratorProducts/interface/HepMCProduct.h"
+#include "SimDataFormats/GeneratorProducts/interface/GenEventInfoProduct.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "FWCore/Utilities/interface/RandomNumberGenerator.h"
+
+#include "CLHEP/Random/RandFlat.h"
+
+#include "TFile.h"
+
+namespace CLHEP {
+  class HepRandomEngine;
+}
+
+namespace edm {
+  BeamMomentumGunProducer::BeamMomentumGunProducer(const edm::ParameterSet& pset)
+      : FlatBaseThetaGunProducer(pset),
+        parPDGId_(nullptr),
+        parX_(nullptr),
+        parY_(nullptr),
+        parZ_(nullptr),
+        parPx_(nullptr),
+        parPy_(nullptr),
+        parPz_(nullptr),
+        b_npar_(nullptr),
+        b_eventId_(nullptr),
+        b_parPDGId_(nullptr),
+        b_parX_(nullptr),
+        b_parY_(nullptr),
+        b_parZ_(nullptr),
+        b_parPx_(nullptr),
+        b_parPy_(nullptr),
+        b_parPz_(nullptr) {
+    edm::ParameterSet pgun_params = pset.getParameter<edm::ParameterSet>("PGunParameters");
+
+    // doesn't seem necessary to check if pset is empty - if this
+    // is the case, default values will be taken for params
+    zpos_ = pgun_params.getParameter<double>("ZPosition");
+    edm::FileInPath fp = pgun_params.getParameter<edm::FileInPath>("FileName");
+    std::string infileName = fp.fullPath();
+
+    fFile_ = new TFile(infileName.c_str());
+    fFile_->GetObject("EventTree", fTree_);
+    nentries_ = fTree_->GetEntriesFast();
+    if (fVerbosity > 0)
+      edm::LogVerbatim("BeamMomentumGun") << "Total Events: " << nentries_ << " in " << infileName << " Z " << zpos_;
+
+    // Set branch addresses and branch pointers
+    fTree_->SetBranchAddress("npar", &npar_, &b_npar_);
```

Comment:
for a future PR: it would make this code safer to check the status of each branch (whether it exists in the tree and is active), in case someone tries to run this with an incorrectly formatted input tree